### PR TITLE
feat: new UserSession tied to a non-federated ArcGIS Server instance

### PR DESCRIPTION
**Pull request by jgravois**
the proposed tweak to `UserSession` would enable developers to build up a UI of their own for users to enter credentials and access secure content from unfederated instances of ArcGIS Server.

```js
import { request } from "@esri/arcgis-rest-request";
import { UserSession } from "@esri/arcgis-rest-auth";

const authentication = new UserSession({
  username: "user1",
  password: "user1",
  server: "https://sampleserver6.arcgisonline.com/arcgis"
});

request('https://sampleserver6.arcgisonline.com/arcgis/rest/services/Wildfire_secure/MapServer', { authentication })
  .then(token)
```

AFFECTS PACKAGES:
@esri/arcgis-rest-auth

ISSUES CLOSED: #174

### CODE REVIEW

File: packages/arcgis-rest-auth/src/UserSession.ts [hunk 1]
```diff
@@ -323,8 +333,10 @@ export class UserSession implements IAuthenticationManager {
     this.tokenDuration = options.tokenDuration || 20160;
     this.redirectUri = options.redirectUri;
     this.refreshTokenTTL = options.refreshTokenTTL || 1440;
-
     this.trustedServers = {};
+    if (options.server) {
+      this.trustedServers[options.server] = null;
```
Comment (tomwayson) [inline packages/arcgis-rest-auth/src/UserSession.ts]: I _think_ I know why you're doing this, but probably a good idea to leave a comment here.

File: packages/arcgis-rest-auth/src/UserSession.ts [hunk 2]
```diff
@@ -760,7 +790,8 @@ export class UserSession implements IAuthenticationManager {
             params: {
               username: this.username,
               password: this.password,
-              expiration: this.tokenDuration
+              expiration: this.tokenDuration,
+              client: "referer"
```
Comment (tomwayson) [inline packages/arcgis-rest-auth/src/UserSession.ts]: Do we _always_ want to send `client=referer`? Do we need to add that to above call to `generateToken()` too?

File: packages/arcgis-rest-auth/src/UserSession.ts [hunk 3]
```diff
@@ -722,25 +734,43 @@ export class UserSession implements IAuthenticationManager {
...
+           * be able to generate a token
+           */
+          if (!new RegExp(response.owningSystemUrl, "i").test(this.portal)) {
+            throw new ArcGISAuthError(
+              `${url} is not federated with ${this.portal}.`,
+              "NOT_FEDERATED"
+            );
+          } else {
+            return request(
+              `${response.owningSystemUrl}/sharing/rest/info`,
```
Comment (tomwayson) [inline packages/arcgis-rest-auth/src/UserSession.ts]: Maybe add a comment here to 'splain what this case is?

File: packages/arcgis-rest-auth/src/UserSession.ts [hunk 4]
```diff
@@ -722,25 +734,43 @@ export class UserSession implements IAuthenticationManager {
...
+          }
+        } else if (
+          response.authInfo &&
+          this.trustedServers[root] !== undefined
         ) {
+          /**
+           * if its a stand-alone instance of ArcGIS Server that doesn't advertise
+           * federation at all and the root url is recognized, use its built in token endpoint.
+           */
+          return new Promise((resolve, reject) => {
```
Comment (tomwayson) [inline packages/arcgis-rest-auth/src/UserSession.ts]: Maybe just `return Promise.resolve({ authInfo: response.authInfo });`

**Comment by tomwayson** [comment]
@jgravois

I believe we'd also need to make further changes if we wanted to support @dbouwman's [idea](https://github.com/Esri/arcgis-rest-js/issues/422#issuecomment-448778878)

> where the consuming app has the token (from a cookie, localstorage, some other auth flow managed by the app), but not the username/password?

Maybe set the token here?

https://github.com/Esri/arcgis-rest-js/pull/423#discussion-diff-243449223R338

File: packages/arcgis-rest-auth/src/UserSession.ts (hunk 2 above)
Comment (jgravois) [reply packages/arcgis-rest-auth/src/UserSession.ts]: it was only _necessary_ in the case that i tweaked, but there's no harm in being consistent.

File: packages/arcgis-rest-auth/src/UserSession.ts (hunk 4 above)
Comment (jgravois) [reply packages/arcgis-rest-auth/src/UserSession.ts]: 💡 	
👶

**Comment by jgravois** [comment]
> where the consuming app has the token (from a cookie, localstorage, some other auth flow managed by the app), but not the username/password?

i've added another small tweak to support the use case you mentioned. this is what needs to be provided to invoke it:

```js
const authentication = new UserSession({
  server: "https://sampleserver6.arcgisonline.com/arcgis",
  token: "abc123",
  tokenExpires: new Date(1545415669763)
});
```

to do:
- [ ] get @patrickarlt's blessing
- [ ] refactor the trusted server check to compare only root urls

**Comment by jgravois** [comment]
@patrickarlt given that you chimed in on #422 but not this PR i'll _assume_ you aren't **fundamentally opposed** to providing a mechanism to support making requests to non-federated, authenticated services.

i *know* you're busy with https://developers.arcgis.com tasks so if i'm wrong even a quick "hold your horses, i haven't had time to look at this yet and i'd like to", sometime this week would be welcome.

**Comment by patrickarlt** [comment]
@jgravois I realized that you asked about this again ago but it all looks good to me. I figured at some point someone would ask for this, I just didn't want to do it.

**Comment by jgravois** [comment]
thanks @patrickarlt!

i've rebased and pushed up some more changes to ensure test coverage and tie up the remaining loose ends. i'm happy with this if @tomwayson is.

**Pull request merged**